### PR TITLE
[tensor-shapes] add stubs for jax.numpy array structure APIs

### PR DESCRIPTION
**Pull request by jakevdp**
### Summary

Type stubs for jax structural ops: concatenate, stack, split, dimension operations, etc.

### Test Plan

Adds a new test file; all tests pass locally:
```
$ uv tool run --from ruff==0.14.3 ruff format --check --config ruff.toml .
328 files already formatted
$ uv tool run --from ruff==0.14.3 ruff check --select I --config ruff.toml .
All checks passed!
$ uv run python test.py --no-test --no-conformance --no-jsonschema
Running formatting...
Finished in 2.68 seconds.
Running linting...
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.87s
Finished in 1.08 seconds.
$ uv run python tensor-shapes/pyrefly-jax-stubs/run_pyrefly.py --release
+ cargo build -p pyrefly --release
    Finished `release` profile [optimized + debuginfo] target(s) in 0.56s
PASS stubs (9 files)
PASS arithmetic (1 files)
PASS creation (1 files)
PASS fft (1 files)
PASS lax (1 files)
PASS linalg (1 files)
PASS matmul (1 files)
PASS nn (1 files)
PASS reductions (1 files)
PASS reshape (1 files)
PASS structural (1 files)
```

### AI disclosure

Change generated with a Gemini coding agent.

### CODE REVIEW

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi [hunk 1]
```diff
@@ -1056,6 +1094,179 @@ def reshape(
...
+def flip[Shape: _Shape](
+    m: Array[Shape],
+    axis: Sequence[int] | None = None,
+) -> Array[Shape]: ...
+def fliplr[Shape: _Shape](
+    m: Array[Shape],
+) -> Array[fliplr_shape(Shape)]: ...
+def flipud[Shape: _Shape](
+    m: Array[Shape],
+) -> Array[flipud_shape(Shape)]: ...
```
Comment (jakevdp) [inline tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi]: One question on these: for both of these the output shape is equal to the input shape, but there is a lower-bound on the number of dimensions (`fliplr` requires an array with at least one dimension, `flipud` requires at least two).

We could also express this without a DSL function, via something like `Array[[*Elements[Batch], N]]` and `Array[[*Elements[Batch], N, M]]`.

Which approach is preferable?

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi [hunk 2]
```diff
@@ -1056,6 +1094,179 @@ def reshape(
...
+@overload
+def roll[Shape: _Shape, Axis: Flag[_Axis] = None](
+    a: Array[Shape],
+    shift: Any,
+    axis: Axis = None,
+) -> Array[roll_shape(Shape, Axis)]: ...
```
Comment (jakevdp) [inline tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi:1460]: Notice that for `roll` and `flip`, the output shape always equals the input shape. The only reason to have a DSL function is to validate that `axis` is in-bounds.

Should we be doing this kind of validation, or should we just focus on shape propagation, avoiding the DSL call?

File: tensor-shapes/test-requirements.txt [hunk 3]
```diff
@@ -18,9 +18,9 @@
 --extra-index-url https://download.pytorch.org/whl/cpu
 
 numpy==2.5.0
-jax==0.10.2
+jax==0.11.1
 # Pinned alongside `jax`: it is a separate distribution that would otherwise
 # resolve freely under a pinned `jax`.
-jaxlib==0.10.2
+jaxlib==0.11.1
```
Comment (jakevdp) [inline tensor-shapes/test-requirements.txt:24]: This version bump was required for the fix in https://github.com/jax-ml/jax/pull/39075, which properly supports negative `axis` in `jnp.permute_dims`.

**Review by stroxler — APPROVED** [review]
Review automatically exported from Phabricator review in Meta.

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi (hunk 1 above)
Comment (stroxler) [reply tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi]: I'm honestly not sure - I think `Array[[*Elements[Batch], N, M]]` is probably clearer for someone actually reading the types, whereas the dsl function allows cleaner error messages

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi (hunk 2 above)
Comment (stroxler) [reply tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi:1460]: I think using the DSL for validation is fine, we already have some similar examples in torch stubs (for example `cosine_similarity_shape` is entirely validation)

File: tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi (hunk 1 above)
Comment (jakevdp) [reply tensor-shapes/pyrefly-jax-stubs/jax-stubs/numpy/__init__.pyi]: OK, I'll make that switch.
